flasksrc/session.py

```diff
@@ -556,6 +575,234 @@ def put(self, toynet_session_id, **kwargs):
+                    break
+
+            # We select the first available interface on the router
+            if available_intf:
+                dev1_intf.text = str(intf)
+                break
+        else:  # Refers to the outer for loop, not the if
+            # No available interfaces
+            abort(400, message='Device {req["dev_1"]} has no available interfaces')
+
+        dev2_intf.text = self.getNextIntfForSwitch(root, dev2.attrib['name'])
+
+        dev1.append(dev1_intf)
+        dev2.append(dev2_intf)
+        link.append(dev1)
+        link.append(dev2)
+
+        # Update topology with changes
+        root.find('linkList').append(link)
+        return ET.tostring(root).decode('utf-8')
+
+    def linkSwitchSwitch(self, dev1_elem, dev2_elem, root, req):
+        link = ET.Element('link')
+        dev1 = ET.Element('dvc')
+        dev2 = ET.Element('dvc')
+        dev1_intf = ET.Element('intf')
+        dev2_intf = ET.Element('intf')
+
+        # Create new interfaces for both switches
+        dev1_intf.text = self.getNextIntfForSwitch(root, req['dev_1'])
+        dev2_intf.text = self.getNextIntfForSwitch(root, req['dev_2'])
+        dev1.attrib['name'] = req['dev_1']
+        dev2.attrib['name'] = req['dev_2']
+
+        dev1.append(dev1_intf)
+        dev2.append(dev2_intf)
+        link.append(dev1)
+        link.append(dev2)
+
+        # Update topology with changes
+        root.find('linkList').append(link)
+        return ET.tostring(root).decode('utf-8')
+
+    def linkHostSwitch(self, dev1_elem, dev2_elem, root, req):
+        link = ET.Element('link')
+        dev1 = ET.Element('dvc')
+        dev2 = ET.Element('dvc')
+        dev2_intf = ET.Element('intf')
+
+        # Always make the host dev1 for simplicity
+        if dev1_elem.tag != 'host':
+            dev2.attrib['name'] = req['dev_1']
+            dev1.attrib['name'] = req['dev_2']
+            dev1_elem, dev2_elem = dev2_elem, dev1_elem
+        else:
+            dev1.attrib['name'] = req['dev_1']
+            dev2.attrib['name'] = req['dev_2']
+
+        # No host interface specification needed, just the switch
+        dev2_intf.text = self.getNextIntfForSwitch(root, dev2.attrib['name'])
+
+        dev2.append(dev2_intf)
+        link.append(dev1)
+        link.append(dev2)
+
+        # Update topology with changes
+        root.find('linkList').append(link)
+        return ET.tostring(root).decode('utf-8')
+
+    def extractEndpoints(self, root, req):
+        dev1_elem = None
+        dev2_elem = None
+
+        # Find the specified devices
+        dev_list = root.find('routerList').findall('router')
+        dev_list += root.find('switchList').findall('switch')
+        dev_list += root.find('hostList').findall('host')
+        for device in dev_list:
+            if device.get('name') == req['dev_1']:
+                dev1_elem = device
+            if device.get('name') == req['dev_2']:
+                dev2_elem = device
```

Comment:
We use `from` and `to` to indicate a link in the frontend, not sure how you feel about that nomenclature though.

```suggestion
            if device.get('name') == req['from']:
                dev1_elem = device
            if device.get('name') == req['to']:
                dev2_elem = device
```

Reply:
I chose this nomenclature since links are undirected; if `to` and `from` are prevalent in the frontend I'm open to changing them here to keep things simpler though.